core/server/common/src/main/java/alluxio/master/journal/raft/transport/CopycatGrpcConnection.java

```diff
@@ -0,0 +1,439 @@
+        Throwable error = mContext.serializer().readObject(response.getMessage().newInput());
+        LOG.debug("Received error response for request: {}. Error: {}",
+            response.getResponseHeader().getRequestId(), error);
+        future.getContext().executor().execute(() -> future.completeExceptionally(error));
+      } else {
+        Object responseObject = mContext.serializer().readObject(response.getMessage().newInput());
+        LOG.debug("Received response of type: {} request: {}.", responseObject.getClass().getName(),
+            response.getResponseHeader().getRequestId());
+        future.getContext().executor().execute(() -> future
+            .complete(responseObject));
+      }
+    } catch (SerializationException e) {
+      future.getContext().executor().execute(() -> future.completeExceptionally(e));
+    }
+  }
+
+  @Override
+  public Listener<Throwable> onException(Consumer<Throwable> listener) {
+    // Call immediately if the connection was faulted.
+    if (mLastFailure != null) {
+      listener.accept(mLastFailure);
+    }
+    return mExceptionListeners.add(Assert.notNull(listener, "listener"));
+  }
+
+  @Override
+  public Listener<Connection> onClose(Consumer<Connection> listener) {
+    // Call immediately if the connection was closed.
+    if (mClosed) {
+      listener.accept(this);
+    }
+    return mCloseListeners.add(Assert.notNull(listener, "listener"));
+  }
+
+  @Override
+  public CompletableFuture<Void> close() {
+    if (!mClosed) {
+      LOG.debug("Closing connection. Owner: {}", mConnectionOwner);
+      mTargetObserver.onCompleted();
+
+      mClosed = true;
+      // Stop timeout timer.
+      mTimeoutScheduler.cancel();
+
+      // Close outstanding calls.
+      for (CopycatGrpcConnection.ContextualFuture<?> responseFuture : mResponseFutures.values()) {
+        responseFuture.getContext().executor().execute(
+            () -> responseFuture.completeExceptionally(new ConnectException("Connection closed.")));
+      }
+      mResponseFutures.clear();
+
+      // Invoke close listeners.
+      for (Listener<Connection> listener : mCloseListeners) {
+        listener.accept(this);
+      }
+    }
+
+    return CompletableFuture.completedFuture(null);
+  }
+
+  /*
+   * gRPC message handlers.
+   */
+
+  @Override
+  public void onNext(CopycatMessage message) {
+    LOG.debug("Received message with requestHeader: {}, responseHeader: {}",
+        message.getRequestHeader(), message.getResponseHeader());
+    // A message can be a request or a response.
+    if (message.hasRequestHeader()) {
+      handleRequestMessage(message);
+    } else if (message.hasResponseHeader()) {
+      handleResponseMessage(message);
+    } else {
+      throw new RuntimeException("Message should contain a request/response header.");
+    }
+  }
+
+  @Override
+  public void onError(Throwable t) {
+    LOG.debug("Connection faulted. Owner: {}. Error: {}", mConnectionOwner, t);
```

Comment:
(nit) `faulted` -> `failed`?

also you don't need to explicitly print the error as an argument. log4j will print out the full exception that is specified as an extra argument.